File: Hans.Contoso/Hans.Contoso.Web/Scripts/App/CustomerController.js
//'use strict';
(function () {
    var app = angular.module('ContosoRetailDW', ['ui.bootstrap', 'toastr', 'angular-loading-bar'])
        .config(['cfpLoadingBarProvider', function (cfpLoadingBarProvider) {
            cfpLoadingBarProvider.includeSpinner = false;
        }]);

    app.controller('CustomerController', function ($scope, $http, $modal, toastr) {
        $scope.pagingInfo = {
            pagingSize: 12,
            page: 1,
            pageSize: 10,
            sortBy: 'CustomerKey',
            isAsc: true,
            totalItems: 0
        };

        $scope.gridIndex = function (index) {
            return (index + 1) + (($scope.pagingInfo.page - 1) * $scope.pagingInfo.pageSize);
        };

        $scope.setPage = function (pageNo) {
            $scope.pagingInfo.page = pageNo;
        };

        $scope.pageChanged = function () {
            pageInit();
        };

        $scope.sort = function (sortBy) {
            if (sortBy === $scope.pagingInfo.sortBy) {
                $scope.pagingInfo.isAsc = !$scope.pagingInfo.isAsc;
            } else {
                $scope.pagingInfo.sortBy = sortBy;
                $scope.pagingInfo.isAsc = false;
            }
            pageInit();
        };

        $scope.openModal = function (id, html) {
            var modalInstance = $modal.open({
                templateUrl: html,
                controller: 'ModalController',
                size: '',
                resolve: {
                    selectedID: function () {
                        return id;
                    },
                    selectedPagingInfo: function () {
                        return $scope.pagingInfo;
                    }
                }
            });

            modalInstance.result.then(function (refreshCustomers) {
                $scope.customers = refreshCustomers;
            }, function () {
                //$log.info('Modal dismissed at: ' + new Date());
            });
        };

        function loadTotalItems() {
            $http.get('/api/customer/getsize')
            .success(function (data, status, headers, config) {
                $scope.pagingInfo.totalItems = data;
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.error($scope.error);
            });
        }

        function loadCustomers() {
            $http.get('/api/customer/getallby', { params: $scope.pagingInfo })
            .success(function (data, status, headers, config) {
                $scope.customers = data;
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.error($scope.error);
            });
        }

        function pageInit() {
            loadTotalItems();
            loadCustomers();
            toastr.info('Load page ' + $scope.pagingInfo.page + ' and sort by ' + $scope.pagingInfo.sortBy);
        }

        pageInit();
    });

    app.controller('ModalController', function ($scope, $modalInstance, $http, selectedID, selectedPagingInfo) {
        $scope.selectedID = selectedID;
        $scope.selectedPagingInfo = selectedPagingInfo;

        $scope.ok = function () {
            closeAndRefreshRepeater();
        };

        $scope.cancel = function () {
            closeAndRefreshRepeater();
        };

        $scope.create = function (customer) {
            if ($scope.createForm.$valid) {
                $scope.isProcessing = true;
                $http({
                    method: 'POST',
                    url: '/api/customer/create',
                    data: {
                        'CustomerLabel': customer.CustomerLabel,
                        'Title': customer.Title,
                        'FirstName': customer.FirstName,
                        'MiddleName': customer.MiddleName,
                        'LastName': customer.LastName,
                        'NameStyle': customer.NameStyle,
                        'BirthDate': customer.BirthDate,
                        'MaritalStatus': customer.MaritalStatus,
                        'Suffix': customer.Suffix,
                        'Gender': customer.Gender,
                        'EmailAddress': customer.EmailAddress,
                        'YearlyIncome': customer.YearlyIncome,
                        'TotalChildren': customer.TotalChildren,
                        'NumberChildrenAtHome': customer.NumberChildrenAtHome,
                        'Education': customer.Education,
                        'Occupation': customer.Occupation,
                        'HouseOwnerFlag': customer.HouseOwnerFlag,
                        'NumberCarsOwned': customer.NumberCarsOwned,
                        'AddressLine1': customer.AddressLine1,
                        'AddressLine2': customer.AddressLine2,
                        'Phone': customer.Phone,
                        'DateFirstPurchase': customer.DateFirstPurchase,
                        'CustomerType': customer.CustomerType,
                        'CompanyName': customer.CompanyName,
                        'ETLLoadID': customer.ETLLoadID,
                        'LoadDate': customer.LoadDate,
                        'UpdateDate': customer.UpdateDate,
                    },
                    headers: { 'Content-Type': 'application/json' }
                })
                .success(function (data, status, headers, config) {
                    $scope.isProcessing = false;
                    toastr.pop('success', 'Create successful...');
                    closeAndRefreshRepeater();
                })
                .error(function (data, status, headers, config) {
                    $scope.error = 'Error has occured!';
                    toastr.pop('error', $scope.error);
                });
            }
        };

        $scope.update = function (customer) {
            if ($scope.editForm.$valid) {
                $scope.isProcessing = true;
                $http({
                    method: 'PUT',
                    url: '/api/customer/edit/' + customer.CustomerKey,
                    data: {
                        'CustomerKey': customer.CustomerKey,
                        'GeographyKey': customer.GeographyKey,
                        'CustomerLabel': customer.CustomerLabel,
                        'Title': customer.Title,
                        'FirstName': customer.FirstName,
                        'MiddleName': customer.MiddleName,
                        'LastName': customer.LastName,
                        'NameStyle': customer.NameStyle,
                        'BirthDate': customer.BirthDate,
                        'MaritalStatus': customer.MaritalStatus,
                        'Suffix': customer.Suffix,
                        'Gender': customer.Gender,
                        'EmailAddress': customer.EmailAddress,
                        'YearlyIncome': customer.YearlyIncome,
                        'TotalChildren': customer.TotalChildren,
                        'NumberChildrenAtHome': customer.NumberChildrenAtHome,
                        'Education': customer.Education,
                        'Occupation': customer.Occupation,
                        'HouseOwnerFlag': customer.HouseOwnerFlag,
                        'NumberCarsOwned': customer.NumberCarsOwned,
                        'AddressLine1': customer.AddressLine1,
                        'AddressLine2': customer.AddressLine2,
                        'Phone': customer.Phone,
                        'DateFirstPurchase': customer.DateFirstPurchase,
                        'CustomerType': customer.CustomerType,
                        'CompanyName': customer.CompanyName,
                        'ETLLoadID': customer.ETLLoadID,
                        'LoadDate': customer.LoadDate,
                        'UpdateDate': customer.UpdateDate,
                    },
                    headers: { 'Content-Type': 'application/json' }
                })
                .success(function (data, status, headers, config) {
                    $scope.isProcessing = false;
                    toastr.pop('success', 'Update successful...');
                    closeAndRefreshRepeater();
                })
                .error(function (data, status, headers, config) {
                    $scope.error = 'Error has occured!';
                    toastr.pop('error', $scope.error);
                });
            }
        };

        $scope.delete = function (id) {
            $scope.isProcessing = true;
            $http.delete('/api/customer/delete', {
                params: { 'id': id }
            })
            .success(function (data, status, headers, config) {
                $scope.isProcessing = false;
                toastr.pop('success', 'Delete successful...');
                closeAndRefreshRepeater();
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.pop('error', $scope.error);
            });
        };

        function getAvailableGeographies() {
            $scope.availableDimGeography = [];
            $http.get('/api/geography/getall')
            .success(function (data, status, headers, config) {
                $.each(data, function (index, item) {
                    $scope.availableDimGeography.push({ ID: item.GeographyKey, Name: item.GeographyKey });
                });
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.pop('error', $scope.error);
            });
        }

        function loadCustomer() {
            $scope.customer = null;
            $http.get('/api/customer/get', { params: { id: $scope.selectedID } })
            .success(function (data, status, headers, config) {
                $scope.customer = data;
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.pop('error', $scope.error);
            });
        }

        function loadTotalItems() {
            $http.get('/api/customer/getsize')
            .success(function (data, status, headers, config) {
                $scope.selectedPagingInfo.totalItems = data;
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.pop('error', $scope.error);
            });
        }

        function loadCustomers() {
            $http.get('/api/customer/getallby', { params: $scope.selectedPagingInfo })
            .success(function (data, status, headers, config) {
                $scope.refreshCustomers = data;
                $modalInstance.close(data);
            })
            .error(function (data, status, headers, config) {
                $scope.error = 'Error has occured!';
                toastr.pop('error', $scope.error);
            });
        }

        function closeAndRefreshRepeater() {
            toastr.pop('info', 'Unload modal...');
            loadTotalItems();
            loadCustomers();
        }

        function pageInit() {
            toastr.pop('info', 'Load modal...');
            getAvailableGeographies();
            if ($scope.selectedID != '')
                loadCustomer();
        }

        pageInit();
    });

})();

File: Hans.Contoso/Hans.Contoso.Web/Controllers/CustomerController.cs
using Hans.Contoso.Core.Domains;
using Hans.Contoso.Core.Persistence;
using Hans.Contoso.Web.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using System.Web.Http;

namespace Hans.Contoso.Web.Controllers
{
    /// <summary>
    /// Customer API
    /// </summary>
    public class CustomerController : ApiController
    {
        /// <summary>
        /// Customer repository
        /// </summary>
        public IRepository<Customer> CustomerRepository { get; set; }

        /// <summary>
        /// Get customer count
        /// </summary>
        /// <returns>Count of all customers</returns>
        [HttpGet]
        public async Task<int> GetSize()
        {
            var products = await CustomerRepository.FindAllAsync();

            return products.Count();
        }

        /// <summary>
        /// Get all customers
        /// </summary>
        /// <returns>List of all customers</returns>
        [HttpGet]
        public async Task<IQueryable<CustomerModel>> GetAll()
        {
            var list = await CustomerRepository.FindAllAsync();

            return list.Select(x => new CustomerModel
            {
                CustomerKey = x.CustomerKey,
                GeographyKey = x.Geography.GeographyKey,
                CustomerLabel = x.CustomerLabel,
                Title = x.Title,
                FirstName = x.FirstName,
                MiddleName = x.MiddleName,
                LastName = x.LastName,
                NameStyle = x.NameStyle.HasValue ? x.NameStyle.Value : false,
                BirthDate = x.BirthDate.HasValue ? x.BirthDate.Value : DateTime.MinValue,
                MaritalStatus = x.MaritalStatus,
                Suffix = x.Suffix,
                Gender = x.Gender,
                EmailAddress = x.EmailAddress,
                YearlyIncome = x.YearlyIncome.HasValue ? x.YearlyIncome.Value : decimal.MinValue,
                TotalChildren = x.TotalChildren.HasValue ? x.TotalChildren.Value : int.MinValue,
                NumberChildrenAtHome = x.NumberChildrenAtHome.HasValue ? x.NumberChildrenAtHome.Value : int.MinValue,
                Education = x.Education,
                Occupation = x.Occupation,
                HouseOwnerFlag = x.HouseOwnerFlag,
                NumberCarsOwned = x.NumberCarsOwned.HasValue ? x.NumberCarsOwned.Value : int.MinValue,
                AddressLine1 = x.AddressLine1,
                AddressLine2 = x.AddressLine2,
                Phone = x.Phone,
                DateFirstPurchase = x.DateFirstPurchase.HasValue ? x.DateFirstPurchase.Value : DateTime.MinValue,
                CustomerType = x.CustomerType,
                CompanyName = x.CompanyName,
                ETLLoadID = x.ETLLoadID.HasValue ? x.ETLLoadID.Value : int.MinValue,
                LoadDate = x.LoadDate.HasValue ? x.LoadDate.Value : DateTime.MinValue,
                UpdateDate = x.UpdateDate.HasValue ? x.UpdateDate.Value : DateTime.MinValue
            });
        }

        /// <summary>
        /// Get all customer filter by search
        /// </summary>
        /// <param name="search">string</param>
        /// <returns>List of customer</returns>
        [HttpGet]
        public async Task<IQueryable<CustomerModel>> GetAllBy(string search)
        {
            var list = await CustomerRepository.FindAllAsync();

            return list.Select(x => new CustomerModel
            {
                CustomerKey = x.CustomerKey,
                GeographyKey = x.Geography.GeographyKey,
                CustomerLabel = x.CustomerLabel,
                Title = x.Title,
                FirstName = x.FirstName,
                MiddleName = x.MiddleName,
                LastName = x.LastName,
                NameStyle = x.NameStyle.HasValue ? x.NameStyle.Value : false,
                BirthDate = x.BirthDate.HasValue ? x.BirthDate.Value : DateTime.MinValue,
                MaritalStatus = x.MaritalStatus,
                Suffix = x.Suffix,
                Gender = x.Gender,
                EmailAddress = x.EmailAddress,
                YearlyIncome = x.YearlyIncome.HasValue ? x.YearlyIncome.Value : decimal.MinValue,
                TotalChildren = x.TotalChildren.HasValue ? x.TotalChildren.Value : int.MinValue,
                NumberChildrenAtHome = x.NumberChildrenAtHome.HasValue ? x.NumberChildrenAtHome.Value : int.MinValue,
                Education = x.Education,
                Occupation = x.Occupation,
                HouseOwnerFlag = x.HouseOwnerFlag,
                NumberCarsOwned = x.NumberCarsOwned.HasValue ? x.NumberCarsOwned.Value : int.MinValue,
                AddressLine1 = x.AddressLine1,
                AddressLine2 = x.AddressLine2,
                Phone = x.Phone,
                DateFirstPurchase = x.DateFirstPurchase.HasValue ? x.DateFirstPurchase.Value : DateTime.MinValue,
                CustomerType = x.CustomerType,
                CompanyName = x.CompanyName,
                ETLLoadID = x.ETLLoadID.HasValue ? x.ETLLoadID.Value : int.MinValue,
                LoadDate = x.LoadDate.HasValue ? x.LoadDate.Value : DateTime.MinValue,
                UpdateDate = x.UpdateDate.HasValue ? x.UpdateDate.Value : DateTime.MinValue
            });
        }

        /// <summary>
        /// Get all customer with paging capability
        /// </summary>
        /// <param name="page">int</param>
        /// <param name="pageSize">int</param>
        /// <param name="sortBy">string</param>
        /// <param name="isAsc">true</param>
        /// <returns>List of customer with paging capability</returns>
        public async Task<IQueryable<CustomerModel>> GetAllBy(int page, int pageSize, string sortBy, bool isAsc)
        {
            var customers = await CustomerRepository.FindAllAsync();

            switch (sortBy.ToLower())
            {
                case "customerkey":
                    customers = isAsc ? customers.OrderBy(p => p.CustomerKey) : customers.OrderByDescending(p => p.CustomerKey);
                    break;
                case "geographykey":
                    customers = isAsc ? customers.OrderBy(p => p.Geography.GeographyKey) : customers.OrderByDescending(p => p.Geography.GeographyKey);
                    break;
                case "customerlabel":
                    customers = isAsc ? customers.OrderBy(p => p.CustomerLabel) : customers.OrderByDescending(p => p.CustomerLabel);
                    break;
                case "title":
                    customers = isAsc ? customers.OrderBy(p => p.Title) : customers.OrderByDescending(p => p.Title);
                    break;
                case "firstname":
                    customers = isAsc ? customers.OrderBy(p => p.FirstName) : customers.OrderByDescending(p => p.FirstName);
                    break;
                case "middlename":
                    customers = isAsc ? customers.OrderBy(p => p.MiddleName) : customers.OrderByDescending(p => p.MiddleName);
                    break;
                case "lastname":
                    customers = isAsc ? customers.OrderBy(p => p.LastName) : customers.OrderByDescending(p => p.LastName);
                    break;
                case "namestyle":
                    customers = isAsc ? customers.OrderBy(p => p.NameStyle) : customers.OrderByDescending(p => p.NameStyle);
                    break;
                case "birthdate":
                    customers = isAsc ? customers.OrderBy(p => p.BirthDate) : customers.OrderByDescending(p => p.BirthDate);
                    break;
                case "maritalstatus":
                    customers = isAsc ? customers.OrderBy(p => p.MaritalStatus) : customers.OrderByDescending(p => p.MaritalStatus);
                    break;
                case "suffix":
                    customers = isAsc ? customers.OrderBy(p => p.Suffix) : customers.OrderByDescending(p => p.Suffix);
                    break;
                case "gender":
                    customers = isAsc ? customers.OrderBy(p => p.Gender) : customers.OrderByDescending(p => p.Gender);
                    break;
                case "emailaddress":
                    customers = isAsc ? customers.OrderBy(p => p.EmailAddress) : customers.OrderByDescending(p => p.EmailAddress);
                    break;
                case "yearlyincome":
                    customers = isAsc ? customers.OrderBy(p => p.YearlyIncome) : customers.OrderByDescending(p => p.YearlyIncome);
                    break;
                case "totalchildren":
                    customers = isAsc ? customers.OrderBy(p => p.TotalChildren) : customers.OrderByDescending(p => p.TotalChildren);
                    break;
                case "numberchildrenathome":
                    customers = isAsc ? customers.OrderBy(p => p.NumberChildrenAtHome) : customers.OrderByDescending(p => p.NumberChildrenAtHome);
                    break;
                case "education":
                    customers = isAsc ? customers.OrderBy(p => p.Education) : customers.OrderByDescending(p => p.Education);
                    break;
                case "occupation":
                    customers = isAsc ? customers.OrderBy(p => p.Occupation) : customers.OrderByDescending(p => p.Occupation);
                    break;
                case "houseownerflag":
                    customers = isAsc ? customers.OrderBy(p => p.HouseOwnerFlag) : customers.OrderByDescending(p => p.HouseOwnerFlag);
                    break;
                case "numbercarsowned":
                    customers = isAsc ? customers.OrderBy(p => p.NumberCarsOwned) : customers.OrderByDescending(p => p.NumberCarsOwned);
                    break;
                case "addressline1":
                    customers = isAsc ? customers.OrderBy(p => p.AddressLine1) : customers.OrderByDescending(p => p.AddressLine1);
                    break;
                case "addressline2":
                    customers = isAsc ? customers.OrderBy(p => p.AddressLine2) : customers.OrderByDescending(p => p.AddressLine2);
                    break;
                case "phone":
                    customers = isAsc ? customers.OrderBy(p => p.Phone) : customers.OrderByDescending(p => p.Phone);
                    break;
                case "datefirstpurchase":
                    customers = isAsc ? customers.OrderBy(p => p.DateFirstPurchase) : customers.OrderByDescending(p => p.DateFirstPurchase);
                    break;
                case "customertype":
                    customers = isAsc ? customers.OrderBy(p => p.CustomerType) : customers.OrderByDescending(p => p.CustomerType);
                    break;
                case "companyname":
                    customers = isAsc ? customers.OrderBy(p => p.CompanyName) : customers.OrderByDescending(p => p.CompanyName);
                    break;
                case "etlloadid":
                    customers = isAsc ? customers.OrderBy(p => p.ETLLoadID) : customers.OrderByDescending(p => p.ETLLoadID);
                    break;
                case "loaddate":
                    customers = isAsc ? customers.OrderBy(p => p.LoadDate) : customers.OrderByDescending(p => p.LoadDate);
                    break;
                case "updatedate":
                    customers = isAsc ? customers.OrderBy(p => p.UpdateDate) : customers.OrderByDescending(p => p.UpdateDate);
                    break;
            }

            return customers.Select(x => new CustomerModel
            {
                CustomerKey = x.CustomerKey,
                GeographyKey = x.Geography.GeographyKey,
                CustomerLabel = x.CustomerLabel,
                Title = x.Title,
                FirstName = x.FirstName,
                MiddleName = x.MiddleName,
                LastName = x.LastName,
                NameStyle = x.NameStyle.HasValue ? x.NameStyle.Value : false,
                BirthDate = x.BirthDate.HasValue ? x.BirthDate.Value : DateTime.MinValue,
                MaritalStatus = x.MaritalStatus,
                Suffix = x.Suffix,
                Gender = x.Gender,
                EmailAddress = x.EmailAddress,
                YearlyIncome = x.YearlyIncome.HasValue ? x.YearlyIncome.Value : decimal.MinValue,
                TotalChildren = x.TotalChildren.HasValue ? x.TotalChildren.Value : int.MinValue,
                NumberChildrenAtHome = x.NumberChildrenAtHome.HasValue ? x.NumberChildrenAtHome.Value : int.MinValue,
                Education = x.Education,
                Occupation = x.Occupation,
                HouseOwnerFlag = x.HouseOwnerFlag,
                NumberCarsOwned = x.NumberCarsOwned.HasValue ? x.NumberCarsOwned.Value : int.MinValue,
                AddressLine1 = x.AddressLine1,
                AddressLine2 = x.AddressLine2,
                Phone = x.Phone,
                DateFirstPurchase = x.DateFirstPurchase.HasValue ? x.DateFirstPurchase.Value : DateTime.MinValue,
                CustomerType = x.CustomerType,
                CompanyName = x.CompanyName,
                ETLLoadID = x.ETLLoadID.HasValue ? x.ETLLoadID.Value : int.MinValue,
                LoadDate = x.LoadDate.HasValue ? x.LoadDate.Value : DateTime.MinValue,
                UpdateDate = x.UpdateDate.HasValue ? x.UpdateDate.Value : DateTime.MinValue
            }).Skip((page - 1) * pageSize).Take(pageSize);
        }

        //// GET: api/Customer
        //public IEnumerable<string> Get()
        //{
        //    return new string[] { "value1", "value2" };
        //}

        //// GET: api/Customer/5
        //public string Get(int id)
        //{
        //    return "value";
        //}

        //// POST: api/Customer
        //public void Post([FromBody]string value)
        //{
        //}

        //// PUT: api/Customer/5
        //public void Put(int id, [FromBody]string value)
        //{
        //}

        //// DELETE: api/Customer/5
        //public void Delete(int id)
        //{
        //}
    }
}
